### PR TITLE
CRM: WooSync order issues and invoices with the fields fee, shipping_tax, subtotals and the calculated amounts

### PR DESCRIPTION
**Pull request by cleacos**
Several fixes.

## Proposed changes:

This PR fixes multiple issues related to invoices:

- The [fee value from WooCommerce orders is not added to the invoice](https://github.com/Automattic/zero-bs-crm/issues/3017), provoking and difference between the total amount in the order and the payment object and the invoice.
- [The shipping tax from WooCommerce is not added to the invoice](https://github.com/Automattic/zero-bs-crm/issues/3023), provoking a difference between total amounts.
- [The subtotal amount from the WooCommerce order is not set to the invoice](https://github.com/Automattic/zero-bs-crm/issues/3017)
- If you update the invoice, [the shipping tax selected is removed, and the invoice is recalculated with incorrect data](https://github.com/Automattic/zero-bs-crm/issues/2923).

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:
- Install WooCommerce and load at least one product. Setting: add one tax `admin.php?page=wc-settings&tab=tax&section=standard`. Check Shipping in the tax you created.
- Enable Invoicing Settings on JPCRM: tax, discounts, shipping `admin.php?page=zerobscrm-plugin-settings&tab=invbuilder`
- WooSync Setting, active: create invoices from WooCommerce Orders `admin.php?page=zerobscrm-plugin-settings&tab=woosync`
- Create a WooCommerce order, add a product, add a fee, add a tax, add a shipping cost
image

- Recalculate the order if it's needed
- Save it
- Check the JPCRM invoice generated

**In trunk:** the amount will be different because of the fee, and the shipping tax values are absent. The calculated taxes is wrong because of the discount. In the  Preview or PDF invoice file, the subtotal value is `0.00`:

image

image

**In this branch:** All the amounts should match:

image

- Play by adding different items to the orders, like coupons. After updating the order, the invoice amounts should match the order amounts.

### CODE REVIEW

**Comment by gogdzl** [comment]
I made an order and got the wrong value.

This is my WooOrder:

image

I got this in `fix/crm-2885-woosync-fee-field-is-not-added-to-invoice`

image

It looks like WooCommerce treats that shipping tax as a monetary value (dollars), and we are treating it as a percentage of the shipping. Doing the math, `$4.50 product tax + $20 * 10%` -> `$4.50 product tax + $2` -> `$6.50`. But this will have to be better analyzed, as WooCommerce may treat this also as a percentage for some orders based on some parameter.


Interestingly `trunk` gave me the correct total value, despite getting the values for each item wrong:

image

**Comment by cleacos** [comment]
Thank you for the review @gogdzl. If you click on `Recalculate`, in the WooCommerce order, it should recalculate the taxes for the Shipping, using percentage.

**Comment by cleacos** [comment]
I found another issue, it seems that the fee is duplicated if you add more items. Moving back the issue to in progress

**Review by coder-karen — COMMENTED** [review]
Thanks for looking at this!!

Running through some tests I did whilst looking at a similar issue, If I’ve modified the tax rates on the CRM tax settings page or I have more than one with the same name, the first is chosen (even if that wasn't the rate used on the order). If I delete the tax rates in the CRM tax settings, then WooSync again, I'd get the right amounts (in the below example that would be a 20% tax).

Order:
Screenshot 2023-04-25 at 08 36 16

Invoice:
Screenshot 2023-04-25 at 08 35 53

But maybe that's for a separate PR / issue?

**Comment by cleacos** [comment]
Thank you for taking a look at it:

> If I’ve modified the tax rates on the CRM tax settings page or I have more than one with the same name, the first is chosen (even if that wasn't the rate used on the order)

Yes, it's a known issue, I don't sure about the current Invoicing system. The invoices should be immutable, and currently, they are mutable, that means that if a tax changes at some point and you update the percentage in the settings, it will change all the old invoices.

**Comment by cleacos** [comment]
So it cannot be resolved in this PR, so at some point, I'd lean toward refactoring it.

**Comment by tbradsha** [comment]
> The invoices should be immutable

🤔 I'm not sure I agree with this (other than invoices from WooSync). But we all agree invoices could use an overhaul. :^)

**Comment by cleacos** [comment]
> I'm not sure I agree with this (other than invoices from WooSync).

It's a standard accounting practice when you issue an invoice, you must not modify it. 

For example, if you have a WooCommerce order with its taxes added, and the order is `Cancelled`, `Refunded`, `Failed`, `Draft`, `Completed` or `Processing`, it doesn't allow you to edit it:

image

Only in these statuses: `Pending payment` or `On hold`, WC allows you to edit the order.

**Review by coder-karen — COMMENTED** [review]
I created an overly complicated order to see what might happen (with taxes, discounts, handling fee, shipping).

This was the Woo order page:
Screenshot 2023-04-28 at 10 26 04

And the CRM invoice after syncing:
Screenshot 2023-04-28 at 10 26 19

It seems to have miscalculated the tax by 1p. Rounding issue?

**Comment by cleacos** [comment]
Yes, probably it's a rounding issue caused on JS calculation side

**Review by coder-karen — APPROVED** [review]
Tests well!

File: projects/plugins/crm/includes/ZeroBSCRM.DAL3.Obj.Invoices.php
Comment (tbradsha) [inline projects/plugins/crm/includes/ZeroBSCRM.DAL3.Obj.Invoices.php:2196]: Interesting...looks like even without defining it it works. But best to explicitly define it like you have in this PR.

File: projects/plugins/crm/js/ZeroBSCRM.admin.invoicebuilder.js
Comment (tbradsha) [inline projects/plugins/crm/js/ZeroBSCRM.admin.invoicebuilder.js:1091]: This is a great example of where `falsey` values are useful (basically "if there's a problem"):

```suggestion
	if ( !v.taxes ) {
```

File: projects/plugins/crm/js/ZeroBSCRM.admin.invoicebuilder.js
Comment (tbradsha) [inline projects/plugins/crm/js/ZeroBSCRM.admin.invoicebuilder.js:1367]:
```suggestion
				tax_percent[ index ] = this_tax_amount / this_row_amount[ index ];
```

File: projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php
Comment (tbradsha) [inline projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php:1216]: Why not always?

File: projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php
Comment (tbradsha) [inline projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php:1728]: Why `false`? In theory this isn't a valid context:
https://github.com/woocommerce/woocommerce/blob/bb244484d38768bc6c33a9423f0a5d5db75976f4/plugins/woocommerce/includes/class-wc-order-item-fee.php#L200

File: projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php
Comment (tbradsha) [inline projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php:1739]: I'm a bit confused here...why would we give a line item a CSV of tax rate IDs instead of a specific tax rate ID?

File: projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php
Comment (tbradsha) [inline projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php:1732]: Why default to `-1` (legacy bad practice) as opposed to not passing it at all?

**Review by tbradsha — COMMENTED** [review]
It seems to work as intended, though I have some questions inline I'd like answered before approval.

File: projects/plugins/crm/includes/ZeroBSCRM.DAL3.Obj.Invoices.php
Comment (cleacos) [reply projects/plugins/crm/includes/ZeroBSCRM.DAL3.Obj.Invoices.php:2196]: Probably in PHP 8, it throws an exception or a warning.

File: projects/plugins/crm/js/ZeroBSCRM.admin.invoicebuilder.js
Comment (cleacos) [reply projects/plugins/crm/js/ZeroBSCRM.admin.invoicebuilder.js:1091]: I'm very rusty with Javascript... but that reduction could work...

File: projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php
Comment (cleacos) [reply projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php:1216]: Because in WooCommerce, you only can recalculate orders while they are in one of those statuses, `pending` or `on-hold`. So we follow that behavior and don't auto-calculate orders in the other statuses.

File: projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php
Comment (cleacos) [reply projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php:1728]: I don't remember exactly what was the issue I had here, but it sounds to me like it was because using the `view` context, it added an additional format, and using `edit` context it returned the raw amount.

File: projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php
Comment (cleacos) [reply projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php:1732]: Yes, I don't like to use -1 either for storing a `null` status. But here it's acting as a simple code, it's not a proper value or an ID. It could have been `9999`.

File: projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php
Comment (cleacos) [reply projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php:1739]: No answer here, legacy code. See line 1694 (old code).

File: projects/plugins/crm/js/ZeroBSCRM.admin.invoicebuilder.js
Comment (tbradsha) [reply projects/plugins/crm/js/ZeroBSCRM.admin.invoicebuilder.js:1091]: Not a blocker. Just something to be aware of.

File: projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php
Comment (tbradsha) [reply projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php:1216]: You're only thinking about the Woo order user interface. But in reality any Woo extension, API connection, etc., could modify the data in the order, so we should probably always recalculate.

File: projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php
Comment (tbradsha) [reply projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php:1732]: Hm...not a blocker, but I'd probably just remove that line and let DAL use whatever the default is (which appears to be an empty string). That way if we ever change to `null` it's one less place to update.

**Review by tbradsha — APPROVED** [review]
No blockers from what I can see, though three suggestions:

* Always recalculate Woo order taxes rather than on specific statuses, as data can change from other sources (Woo API, Woo extensions)
* Don't pass `-1` to the `taxes` line item keys (and update the legacy code) by default, but instead let DAL handle the default
* Figure out why we're added a CSV of taxes to the line item. If we don't know why the legacy code you copied is doing something, and it looks wrong, there's a good chance it is and shouldn't be done again.

File: projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php
Comment (cleacos) [reply projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php:1216]: Believe me, I did it the first time, re-calculate it always, it was the first thing I did, but I got weird behavior (I cannot remember exactly), and that is why I had to make this condition.

File: projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php
Comment (cleacos) [reply projects/plugins/crm/modules/woo-sync/includes/class-woo-sync-background-sync-job.php:1732]: But again, here, consider `-1` as a status code to identify non-taxes. It is used by the `invoicebuilder.js` file to identify those cases via API. This value is not stored in the database. It could have been `0` or `false` too, or just empty. I had to adapt to this legacy code.

I'm not particularly eager to use it, but in this case, I'll leave it as it is.

**Comment by cleacos** [comment]
I did several tests more... Everything works well.

I only found an issue related to the taxes. If you create and remove taxes in WooCommerce and re-add the tax with the same name, it appears the old one in the old invoices but at a 0% rate along with the new tax. It'd be an edge case because this was provoked by the multiple tests and changes I did in the setting taxes page.

**Comment by cleacos** [comment]
> Always recalculate Woo order taxes rather than on specific statuses, as data can change from other sources (Woo API, Woo extensions)

We cannot do it, once we are recalculating orders before syncing invoices. Those two statuses are like a `lock` status. For example, if you change the taxes and orders that are not on that `lock` status, orders will be recalculated, modifying the amounts on sync (JPCRM).

**Comment by cleacos** [comment]
PR Update. Trunk merged.

**Comment by tbradsha** [comment]
> If you create and remove taxes in WooCommerce and re-add the tax with the same name, it appears the old one in the old invoices but at a 0% rate along with the new tax. It'd be an edge case because this was provoked by the multiple tests and changes I did in the setting taxes page.

Did you log this, along with the steps in question, so we can someday fix it?

**Comment by tbradsha** [comment]
> if you change the taxes and orders that are not on that lock status, orders will be recalculated, modifying the amounts on sync (JPCRM).

I don't quite understand this. If you modify the Woo order in any of the "locked" statuses (e.g. via API), are you saying Woo doesn't recalculate numbers?

**Pull request merged**
